programs/checkout/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+use anchor_lang::prelude::*;
+use anchor_spl::{
+    associated_token::AssociatedToken,
+    token::{Mint, Token, TokenAccount},
+};
+
+use solana_program;
+use solana_program::{account_info::AccountInfo, program::invoke, system_instruction};
+
+use exhibition::program::Exhibition;
+use exhibition::{self, Exhibit};
+
+pub mod state;
+use state::heap::Heap;
+use state::linked_list::LinkedList;
+
+declare_id!("8b7yjj2P5fHV9NCyNXJut1pDM1J1D9oRKzqUGW1ycTWk");
+
+#[program]
+pub mod checkout {
+
+    use solana_program::native_token::LAMPORTS_PER_SOL;
+
+    use super::*;
+
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+        msg!("In initializer");
+
+        let list = LinkedList::initialize();
+
+        ctx.accounts.matched_storage.matched_orders = ctx.accounts.matched_orders.key();
+
+        let mut matched_orders = ctx.accounts.matched_orders.load_init()?;
+
+        matched_orders.trades = list;
+
+        msg!("linked holder data {:?}", &matched_orders.trades.order_head);
+        Ok(())
+    }
+
+    pub fn make_bid(ctx: Context<MakeBid>, bid_price: u64) -> Result<()> {
+        msg!("in make bid");
+        msg!("bid price: {}", bid_price / LAMPORTS_PER_SOL);
+
+        let bid_price_sol = bid_price;
+
+        invoke(
+            &system_instruction::transfer(
+                ctx.accounts.bidder.to_account_info().key,
+                ctx.accounts.escrow_sol.to_account_info().key,
+                bid_price_sol,
+            ),
+            &[
+                ctx.accounts.bidder.to_account_info(),
+                ctx.accounts.escrow_sol.to_account_info(),
+                ctx.accounts.system_program.to_account_info(),
+            ],
+        );
+
+        let bid_orders = &mut ctx.accounts.bid_orders.load_mut()?;
+
+        bid_orders
+            .heap
+            .add(bid_price, ctx.accounts.bidder.to_account_info().key());
+
+        let highest_bid = bid_orders.heap.peek_highest_bid();
+
+        msg!("current highest BID : {}", highest_bid.bid_price);
+
+        Ok(())
+    }
+
+    pub fn cancel_bid(ctx: Context<CancelBid>) -> Result<()> {
+        let bidder = &ctx.accounts.bidder;
+
+        let mut heap = ctx.accounts.bid_orders.load_mut()?;
+
+        // Need a clever way to somehow know the bid price after the let mut heap declaration
+        let bid_price_sol = heap.heap.cancel_bid(bidder.key());
```

Comment:
We discussed this, but this function needs to be adjusted to isolate an exact order.